packages/babel-parser/src/plugins/placeholders.js

```diff
@@ -0,0 +1,313 @@
+      );
+    }
+
+    checkLVal(expr: N.Expression): void {
+      if (expr.type !== "Placeholder") super.checkLVal(...arguments);
+    }
+
+    toAssignable(node: N.Node): N.Node {
+      if (
+        node &&
+        node.type === "Placeholder" &&
+        node.expectedNode === "Expression"
+      ) {
+        node.expectedNode = "Pattern";
+        return node;
+      }
+      return super.toAssignable(...arguments);
+    }
+
+    /* ============================================================ *
+     * parser/statement.js                                          *
+     * ============================================================ */
+
+    verifyBreakContinue(node: N.BreakStatement | N.ContinueStatement) {
+      if (node.label && node.label.type === "Placeholder") return;
+      super.verifyBreakContinue(...arguments);
+    }
+
+    parseExpressionStatement(
+      node: MaybePlaceholder<"Statement">,
+      expr: N.Expression,
+    ): MaybePlaceholder<"Statement"> {
+      if (
+        expr.type !== "Placeholder" ||
+        (expr.extra && expr.extra.parenthesized)
+      ) {
+        return super.parseExpressionStatement(...arguments);
+      }
+
+      if (this.match(tt.colon)) {
+        const stmt: N.LabeledStatement = node;
+        stmt.label = this.finishPlaceholder(expr, "Identifier");
+        this.next();
+        stmt.body = this.parseStatement("label");
+        return this.finishNode(stmt, "LabeledStatement");
+      }
+
+      this.semicolon();
+
+      node.name = expr.name;
+      return this.finishPlaceholder(node, "Statement");
+    }
+
+    parseBlock(): MaybePlaceholder<"BlockStatement"> {
+      return (
+        this.parsePlaceholder("BlockStatement") ||
+        super.parseBlock(...arguments)
+      );
+    }
+
+    parseFunctionId(): ?MaybePlaceholder<"Identifier"> {
+      return (
+        this.parsePlaceholder("Identifier") ||
+        super.parseFunctionId(...arguments)
+      );
+    }
+
+    parseClass<T: N.Class>(
+      node: T,
+      isStatement: /* T === ClassDeclaration */ boolean,
+      optionalId?: boolean,
+    ): T {
+      const type = isStatement ? "ClassDeclaration" : "ClassExpression";
+
+      this.next();
+      this.takeDecorators(node);
+
+      const placeholder = this.parsePlaceholder("Identifier");
+      if (placeholder) {
+        if (
+          this.match(tt._extends) ||
+          this.match(tt.placeholder) ||
+          this.match(tt.braceL)
+        ) {
+          node.id = placeholder;
+        } else if (optionalId || !isStatement) {
+          node.id = null;
+          node.body = this.finishPlaceholder(placeholder, "ClassBody");
+          return this.finishNode(node, type);
+        } else {
+          this.unexpected(null, "A class name is required");
+        }
+      } else {
+        this.parseClassId(node, isStatement, optionalId);
+      }
+
+      this.parseClassSuper(node);
+      node.body = this.parsePlaceholder("ClassBody") || this.parseClassBody();
```

Comment:
I think we need to set `constructorAllowsSuper` here too. 
```suggestion
      node.body = this.parsePlaceholder("ClassBody") || this.parseClassBody(!!node.superClass);
```